presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java

```diff
@@ -68,20 +70,26 @@
 
     private final Duration clientTimeout;
 
+    private final ResourceGroupManager<?> resourceGroupManager;
     private final ScheduledExecutorService queryManagementExecutor;
 
     @GuardedBy("this")
     private ScheduledFuture<?> backgroundTask;
 
-    public QueryTracker(QueryManagerConfig queryManagerConfig, ScheduledExecutorService queryManagementExecutor)
+    public QueryTracker(
+            QueryManagerConfig queryManagerConfig,
+            ResourceGroupManager resourceGroupManager,
```

Comment:
This is going to be really tough to fix once we move to a world where resource manager is out of process.  I would prefer that we think of a way of accomplishing this without having to set an atomic boolean on the resource group manager itself.

Reply:
@tdcmeehan Eventually, we should use CPU/Memory/IO as resource budget inside each resource group (currently it is simply query count?).  Before that happens, task count is a proxy to fix deficiency in resource group manager. A global task count limit prevents scheduling a burst of new queries when cluster load is high, similar to total query count limit but more powerful.    The added boolean value `taskLimitExceeded` is light weighted and could be easily deprecated whenever there is better solution.

Reply:
I understand that the flag is lightweight, and it makes sense that we should keep it as a flag to avoid having to deprecate the field once we move to smarter resource placement.  That being said, there doesn't seem to be a good reason to introduce this dependency in `QueryTracker`.  Resource groups already keep track of all running queries and have the full information of the running task count, we just need to access it.  Why not make the decision on whether or not to globally prevent dispatching in `QueryTracker` when we can keep it local to resource groups?  It just seems to confusing to put some of that logic here.

Reply:
I got your point.
-  Traversing resource group tree while taking lock to calculate total number of tasks is more expensive than doing the same thing inside `QueryTracker`. It is duplicated calculation since `QueryTracker` already has that information.
-  The dependency is already there between `SqlQueryManager` and `ResourceGroupManager`,  what you suggested is to migrate some functionalities of `SqlQueryManager` to `ResourceGroupManager`, which is nice.  Given this is how currently it is designed, I still think `QueryTracker` is the best fit and clean implementation.